lib/elixir/src/elixir_clauses.erl

```diff
@@ -46,22 +46,14 @@ guard(Guard, E) ->
   warn_zero_length_guard(EGuard, EG),
   {EGuard, EG}.
 
-warn_zero_length_guard({{'.', _, [erlang, '==']}, Meta,
-                        [{{'.', _, [erlang, length]}, _, [Arg]}, 0]}, E) ->
-  ArgString = 'Elixir.Macro':to_string(Arg),
-  Message = io_lib:format("do not use \"length(~ts) == 0\" to check if a list is empty "
-                          "since length always traverses the whole list. "
-                          "Prefer to pattern match on an empty list or use "
-                          "\"~ts == []\" as a guard", [ArgString, ArgString]),
-  elixir_errors:warn(?line(Meta), ?key(E, file), Message);
-warn_zero_length_guard({{'.', _, [erlang, '>']}, Meta,
-                        [{{'.', _, [erlang, length]}, _, [Arg]}, 0]}, E) ->
-  ArgString = 'Elixir.Macro':to_string(Arg),
-  Message = io_lib:format("do not use \"length(~ts) > 0\" to check if a list is not empty "
-                          "since length always traverses the whole list. "
-                          "Prefer to pattern match on a non-empty list, such as [_ | _], "
-                          "or use \"~ts != []\" as a guard", [ArgString, ArgString]),
-  elixir_errors:warn(?line(Meta), ?key(E, file), Message);
+warn_zero_length_guard({{'.', _, [erlang, Op]}, Meta,
+                        [{{'.', _, [erlang, length]}, _, [Arg]}, 0]}, E) when Op == '=='; Op == '>' ->
+  Warn =
+    case Op of
+      '==' -> {zero_list_length_in_guard, Arg};
+      '>' -> {positive_list_length_in_guard, Arg}
```

Comment:
Looking at this now. Should we also include `>=`? Replacement for that would be a simple `is_list(list)` check (since the only way `lenght(list) >= 0` would fail is when it's not a list).

Reply:
In theory we could also check for `>= 1` (equivalent to `> 0`) and `>= 2` (equivalent to `[_, _ | _]`) and so on. The point is how many of those checks we want to add and how common they are. I think we have a good enough starting point. Not sure if we should add more.